### PR TITLE
SSE implementation of RgbToBgr() function

### PR DESCRIPTION
**Pull request by 0x72D0**
the SSE version is ready to test, only the unit_test for RgbToBgr is missing
relates to #117

### CODE REVIEW

File: src/image_function_simd.cpp [hunk 1]
```diff
@@ -941,6 +942,35 @@ namespace sse
         }
     }
 
+    void RgbToBgr( uint8_t * outY, const uint8_t * inY, const uint8_t * outYEnd, uint32_t rowSizeOut, uint32_t rowSizeIn, uint32_t RGB_simdSize, 
+                   const uint8_t colorCount, uint32_t simdWidth, uint32_t totalSimdWidth, uint32_t nonSimdWidth )
+    {
+        simd ctrl = _mm_set_epi8(15, 12, 13, 14, 9, 10, 11, 6, 7, 8, 3, 4, 5, 0, 1, 2);
```
Comment (ihhub) [inline src/image_function_simd.cpp]: `const simd`

File: src/image_function_simd.cpp [hunk 2]
```diff
@@ -941,6 +942,35 @@ namespace sse
...
+            const uint8_t * inX  = inY;
+            uint8_t       * outX = outY;
+
+            const uint8_t * outXEnd = outX + totalSimdWidth;
+
+            for( ; outX != outXEnd; outX += RGB_simdSize, inX += RGB_simdSize ) {
+                const simd * src = reinterpret_cast<const simd*> (inX);
+                simd * dst = reinterpret_cast<simd*> (outX);
+                simd result = _mm_loadu_si128( src );
+                result = _mm_shuffle_epi8(result, ctrl);
```
Comment (ihhub) [inline src/image_function_simd.cpp]: [MacOS compilation](https://travis-ci.org/ihhub/penguinV/jobs/478124648) fails on this function. Check the bottom of the log.

Comment (ihhub) [reply src/image_function_simd.cpp]: It's probably because this function is SSSE3.

File: src/image_function_simd.cpp [hunk 3]
```diff
@@ -1935,6 +1965,49 @@ if ( simdType == neon_function ) { \
...
+        const uint32_t rowSizeOut = out.rowSize();
+
+        width = width * colorCount;
+
+        const uint8_t * inY  = in.data()  + startYIn  * rowSizeIn  + startXIn  * colorCount;
+        uint8_t       * outY = out.data() + startYOut * rowSizeOut + startXOut * colorCount;
+
+        const uint8_t * outYEnd = outY + height * rowSizeOut;
+
+        const uint32_t RGB_simdSize = (simdSize/3)*3;
```
Comment (ihhub) [inline src/image_function_simd.cpp]: `const uint32_t RGBSimdSize = (simdSize / 3) * 3;`

File: src/image_function_simd.cpp [hunk 4]
```diff
@@ -1935,6 +1965,49 @@ if ( simdType == neon_function ) { \
...
+
+        uint32_t simdWidth = width / RGB_simdSize;
+        uint32_t totalSimdWidth = simdWidth * RGB_simdSize;
+        uint32_t nonSimdWidth = width - totalSimdWidth;
+
+        // to prevent unallowed access to memory
+        if(nonSimdWidth < (simdSize%3))
```
Comment (ihhub) [inline src/image_function_simd.cpp]: if( nonSimdWidth < (simdSize % 3) )

File: src/image_function_simd.cpp (hunk 2 above)
Comment (ihhub) [reply src/image_function_simd.cpp]: This function is declared in `tmmintrin.h` file which is not included here. Check [this](https://software.intel.com/en-us/node/524215)

File: src/image_function_simd.cpp [hunk 5]
```diff
@@ -941,6 +942,35 @@ namespace sse
         }
     }
 
+    void RgbToBgr( uint8_t * outY, const uint8_t * inY, const uint8_t * outYEnd, uint32_t rowSizeOut, uint32_t rowSizeIn, uint32_t RGB_simdSize,
```
Comment (ihhub) [inline src/image_function_simd.cpp]: Use `simdWidth` variable instead of `RGB_simdSize` to keep the same style.

**Comment by 0x72D0** [comment]
ok now we just need on a correct unit_test for this function

**Comment by ihhub** [comment]
I'm working on it right now.

**Comment by ihhub** [comment]
@0x72D0 could you please add performance tests for this pull request while unit test pull request is compiled?

**Comment by ihhub** [comment]
Unit tests are committed. Please merge and add a line to include unit tests for SSE case.

**Comment by 0x72D0** [comment]
the performance is a little bit better

screenshot_20190112_120947

**Comment by ihhub** [comment]
@0x72D0 a small recommendation: please during test run to do execute any other applications giving full CPU usage to tests. Sigma value which comes after average value should be pretty low, <5% than average value. Other applications affect on test results if they eat a lot of CPU resources. If you do such you would have more accurate results so you could check performance enhancement more accurate ;)

**Comment by 0x72D0** [comment]
oh okay I understand, I did it again with only the performance test running
screenshot_20190112_134440

**Comment by ihhub** [comment]
@0x72D0, thank you very much for implementing this!

**Pull request merged**